---
title: "Gene-based test using summary statistics"
author: "Dan Evans"
date: "Last update: `r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    toc: TRUE
---

## Load libraries
```{r}
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(data.table)
library(batchtools)
```

## Set variables
```{r}

pos_offset <- 150000L
gwas_trait <- "fracture" #fracture or bmd, case-sensitive
gwas_dir <- "../data/GWAS/"
output_dir <- "../results/fraxGWAS/"

```

## Overall strategy

Create position annotation object for all gene ids, then use that to annotate the GWAS results.

Then do split/apply/combine by gene id with the GWAS results

## Using TxDb for hg19

One way to extract from db is to retrieve all genes into an object, then subset the object by gene id

`txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene`
`ghs <- genes(txdb)`

extract 1 gene

`ghs[ghs$gene_id == 1,]`

extract multiple genes

`ghs[ghs$gene_id %in% c(1,10,100),]`


## Create position annotation object

Create object with hg19 chr and positions for all genes

```{r, cache = TRUE}
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene

#columns(txdb) #list available meta columns
#query returns all genes
ghs <- as.data.frame(genes(txdb, columns = c("GENEID")), stringsAsFactors = F)
ghs$GENEID <- as.integer(unlist(ghs$GENEID))

#check whether start is always smaller than end, as it should be
if (sum(ghs$start > ghs$end) == 0) print("Start is smaller than end") 
ghs$start <- ghs$start - pos_offset
ghs$end <- ghs$end + pos_offset
ghs$chr <- as.character(ghs$seqnames)
#remove genes not annotated to chromosomes
ghs$chr[grep(".*_.*", ghs$chr)] <- NA
ghs$chr <- gsub("chr", "", ghs$chr)
ghs$chr[ghs$chr %in% c("X")] <- 23
ghs$chr[ghs$chr %in% c("Y")] <- 24
ghs$chr <- as.integer(ghs$chr)
ghs <- ghs[!is.na(ghs$chr),]

```


## Annotate GWAS results with genes based on positions

```{r, cache = TRUE}

if (gwas_trait == "fracture") {
  gwas_file <- "Biobank2-British-FracA-As-C-Gwas-SumStats.txt.gz"
} else {
  gwas_file <- "Biobank2-British-Bmd-As-C-Gwas-SumStats.txt.gz"
}

gwas <- fread(paste0("gunzip -c ", gwas_dir, gwas_file ))
gwas[,gene := -99]

#create iterator for parallel
chunk_size <- 100
if(length(ghs$GENEID) %% chunk_size != 0){
  start_index <- seq(from = 1, to = length(ghs$GENEID), by = chunk_size)
  stop_index <- start_index + chunk_size - 1 
} else {
  start_index <- seq(from = 1, to = length(ghs$GENEID), by = chunk_size)
  stop_index <- start_index + chunk_size - 1 
  stop_index[length(stop_index)] <- length(ghs$GENEID)
}

myindex <- seq_along(start_index)

#annotate GWAS file with gene names in parallel
annotate_gene_pos <- function(iter_index){
  library(data.table)
  for ( i in seq(from = start_index[iter_index], to = stop_index[iter_index], by = 1)){
    mychr <- ghs$chr[i]
    mystart <- ghs$start[i]
    myend <- ghs$end[i]
    mygene <- ghs$GENEID[i]
    if( gwas[CHR == mychr][BP >= mystart & BP <= myend][, .N] > 0 ){
      gwas_sub <- gwas[CHR == mychr & BP >= mystart & BP <= myend, gene := mygene]
    } 
  }
  gwas_sub <- gwas_sub[gene != -99]
  fwrite(gwas_sub, file = paste0(output_dir, "gwas", iter_index, ".csv"))
}


reg <- makeRegistry(file.dir = "myregdir", conf.file = "~/.batchtools.conf.R")
jobs <- myindex
ids <- batchMap(fun = annotate_gene_pos, iter_index = myindex)
batchExport(export = list(ghs = ghs, start_index = start_index, 
			  stop_index = stop_index, gwas = gwas, output_dir = output_dir))
done <- submitJobs(ids, reg = reg, resources = list(partition = "batch", walltime = 180, ntasks = 1, ncpus = 1, memory = 10240))
waitForJobs()
getStatus()
clearRegistry()
removeRegistry(wait = 0, reg = reg)

#concatenate annotated GWAS chunks
all.files <- list.files(path = output_dir, pattern = ".csv")
l <- lapply(all.files, fread)
gwas <- rbindlist(l)

```


## Merge with tables

```{r, eval = FALSE}
gwas[, mean(P.I, na.rm = T), by = gene] # gene-based score



tab <- read.csv("~/bigdata/alliston/HOBeSNP/data/serra_tables/12 TGFb_FDR 0_1.csv", header= T, stringsAsFactors=F)
tab <- read.csv("~/bigdata/alliston/HOBeSNP/data/serra_tables/2 Common DE Aging TGFb.csv", header= T, stringsAsFactors=F)
ghs <- genes(txdb, columns = c("GENEID"),
      filter = list(gene_id = tab$Entrez_Id))
ghs <- as.data.frame(ghs, stringsAsFactors=F)
ghs$GENEID <- as.integer(unlist(ghs$GENEID))
str(ghs)
ghs[1:5,]

geneids <- as.data.frame(org.Hs.egSYMBOL)
#geneids <- select(org.Hs.eg.db, keys = as.character(ghs$GENEID), c("ENTREZID", "GENENAME"), keytype)
ghs <- merge(ghs, geneids, by.x = "GENEID", by.y = "gene_id")
str(ghs)

#there are 343 genes in Serra's table, but only 334 in annotation database
#how many Serra table genes are not found in annotation database?
tab <- merge(tab, ghs, by.x = "Entrez_Id", by.y = "GENEID", all.x = T)
#gene id 199 is on chr 6, but it is absent from annotation file. I'll try a few other files later.


gwas <- fread("gunzip -c ../data/GWAS/Biobank2-British-FracA-As-C-Gwas-SumStats.txt.gz")
gwas <- fread("gunzip -c ../data/GWAS/")
result <- vector(mode = "list", length = length(ghs$GENEID))

#select most significant SNP from each gene region
for ( i in seq_along(ghs$GENEID)){
  mychr <- ghs$chr[i]
  mystart <- ghs$start[i]
  myend <- ghs$end[i]
  if( gwas[CHR == mychr][BP >= mystart & BP <= myend][, .N] > 0 ){
    gwassub <- gwas[CHR == mychr][BP >= mystart & BP <= myend]
    minP <- min(gwassub$P.NI, na.rm=T)
    gwassub2 <- gwassub[P.NI == minP][1]
  } else {
    gwassub <- gwas[1]
    gwassub2 <- as.data.table(lapply(gwassub, function(x) x <- NA))
  }
  result[[i]] <- gwassub2

}
result2 <- rbindlist(result)
ghs <- as.data.table(ghs)
result2 <- cbind(ghs, result2)
fwrite(result2, file = "../results/table12_GWAS_BMD.csv", na = "NA")

```




