---
title: "Gene-based test using summary statistics"
author: "Dan Evans"
date: "Last update: `r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    toc: TRUE
---

```{r, setup}
knitr::opts_chunk$set(cache.lazy = FALSE)
```

## Load libraries
```{r}
library(org.Hs.eg.db)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(data.table)
library(batchtools)
library(qqman)
```

## Set variables
```{r}

pos_offset <- 50000L
gwas_trait <- "fracture" #fracture or bmd, case-sensitive
gwas_dir <- "../data/GWAS/"
output_dir <- "../results/fraxGWAS/"

```

## Overall strategy

Create position annotation object for all gene ids, then use that to annotate the GWAS results.

Then do split/apply/combine by gene id with the GWAS results

## Using TxDb for hg19

One way to extract from db is to retrieve all genes into an object, then subset the object by gene id

`txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene`
`ghs <- genes(txdb)`

extract 1 gene

`ghs[ghs$gene_id == 1,]`

extract multiple genes

`ghs[ghs$gene_id %in% c(1,10,100),]`


## Create position annotation object

Create object with hg19 chr and positions for all genes

```{r}
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene

#columns(txdb) #list available meta columns
#query returns all genes
ghs <- as.data.frame(genes(txdb, columns = c("GENEID")), stringsAsFactors = F)
ghs$GENEID <- as.integer(unlist(ghs$GENEID))

#check whether start is always smaller than end, as it should be
if (sum(ghs$start > ghs$end) == 0) print("Start is smaller than end") 
ghs$start <- ghs$start - pos_offset
ghs$end <- ghs$end + pos_offset
ghs$chr <- as.character(ghs$seqnames)
#remove genes not annotated to chromosomes
ghs$chr[grep(".*_.*", ghs$chr)] <- NA
ghs$chr <- gsub("chr", "", ghs$chr)
ghs$chr[ghs$chr %in% c("X")] <- 23
ghs$chr[ghs$chr %in% c("Y")] <- 24
ghs$chr <- as.integer(ghs$chr)
ghs <- ghs[!is.na(ghs$chr),]

```


## Read in GWAS results and create QQ plot

```{r}

if (gwas_trait == "fracture") {
  gwas_file <- "Biobank2-British-FracA-As-C-Gwas-SumStats.txt.gz"
  gwas <- fread(paste0("gunzip -c ", gwas_dir, gwas_file ))
  gwas[,gene := -99]
  qq(gwas[, P.I], main = "QQ plot fracture")
  mychi <- qchisq(gwas[!is.na(P.I),P.I], 1, lower.tail = FALSE)
  mylambda <- median(mychi, na.rm = T)/qchisq(0.5, 1)
  print(paste0("Lambda = ", mylambda))
} else {
  gwas_file <- "Biobank2-British-Bmd-As-C-Gwas-SumStats.txt.gz"
  gwas <- fread(paste0("gunzip -c ", gwas_dir, gwas_file ))
  gwas[,gene := -99]
  qq(gwas[, P.NI], main = "QQ plot eBMD")
  mychi <- qchisq(gwas[!is.na(P.NI),P.NI], 1, lower.tail = FALSE)
  mylambda <- median(mychi, na.rm = T)/qchisq(0.5, 1)
  print(paste0("Lambda = ", mylambda))
}

```

## Annotate GWAS results with genes based on positions

This code runs in parallel. Below this is code that I set not to evaluate but can run in serial on a laptop.
```{r}
#create iterator for parallel
chunk_size <- 100
if(length(ghs$GENEID) %% chunk_size != 0){
  start_index <- seq(from = 1, to = length(ghs$GENEID), by = chunk_size)
  stop_index <- start_index + chunk_size - 1 
} else {
  start_index <- seq(from = 1, to = length(ghs$GENEID), by = chunk_size)
  stop_index <- start_index + chunk_size - 1 
  stop_index[length(stop_index)] <- length(ghs$GENEID)
}

myindex <- seq_along(start_index)

#annotate GWAS file with gene names in parallel
annotate_gene_pos <- function(iter_index){
  library(data.table)
  for ( i in seq(from = start_index[iter_index], to = stop_index[iter_index], by = 1)){
    gwas_sub <- gwas
    mychr <- ghs$chr[i]
    mystart <- ghs$start[i]
    myend <- ghs$end[i]
    mygene <- ghs$GENEID[i]
    if( gwas_sub[CHR == mychr][BP >= mystart & BP <= myend][, .N] > 0 ){
      gwas_sub[CHR == mychr & BP >= mystart & BP <= myend, gene := mygene]
    } 
  }
  gwas_sub <- gwas_sub[gene != -99]
  fwrite(gwas_sub, file = paste0(output_dir, "gwas", iter_index, ".csv"))
}


reg <- makeRegistry(file.dir = "myregdir", conf.file = "~/.batchtools.conf.R")
jobs <- myindex
ids <- batchMap(fun = annotate_gene_pos, iter_index = myindex)
batchExport(export = list(ghs = ghs, start_index = start_index, 
			  stop_index = stop_index, gwas = gwas, output_dir = output_dir))
done <- submitJobs(ids, reg = reg, resources = list(partition = "batch", walltime = 180, ntasks = 1, ncpus = 1, memory = 10240))
waitForJobs()
getStatus()
clearRegistry()
removeRegistry(wait = 0, reg = reg)


```

Code to run in serial

```{r, eval = FALSE}
#create iterator for serial
chunk_size <- 100
if(length(ghs$GENEID) %% chunk_size != 0){
  start_index <- seq(from = 1, to = length(ghs$GENEID), by = chunk_size)
  stop_index <- start_index + chunk_size - 1 
} else {
  start_index <- seq(from = 1, to = length(ghs$GENEID), by = chunk_size)
  stop_index <- start_index + chunk_size - 1 
  stop_index[length(stop_index)] <- length(ghs$GENEID)
}

myindex <- seq_along(start_index)

#annotate GWAS file with gene names in parallel
annotate_gene_pos <- function(iter_index){
  library(data.table)
  for ( i in seq(from = start_index[iter_index], to = stop_index[iter_index], by = 1)){
    gwas_sub <- gwas
    mychr <- ghs$chr[i]
    mystart <- ghs$start[i]
    myend <- ghs$end[i]
    mygene <- ghs$GENEID[i]
    if( gwas_sub[CHR == mychr][BP >= mystart & BP <= myend][, .N] > 0 ){
      gwas_sub[CHR == mychr & BP >= mystart & BP <= myend, gene := mygene]
    } 
  }
  gwas_sub <- gwas_sub[gene != -99]
  fwrite(gwas_sub, file = paste0(output_dir, "gwas", iter_index, ".csv"))
  return(iter_index)
}

result <- vapply(myindex, annotate_gene_pos, integer(1))


```


## Gene-based scores

Split-apply-combine using data.table by grouping. Do this for each of the chunks to minimize IO strain.

```{r}
library(data.table)
#read chunks of annotated GWAS into a list of DTs
all.files <- list.files(path = output_dir, pattern = "gwas")
l <- lapply(paste0(output_dir, all.files), fread)

#calculate gene-based scores
scores <- function(DT){
  gene_score <- DT[, .(medianG = median(P.I, na.rm = T), minG = min(P.I, na.rm=T)), by = gene]
  return(gene_score)
}

gwas_gene <- lapply(l, scores)
gwas_gene <- rbindlist(gwas_gene)

```

## Merge gene-based scores with gene lists

```{r}

tab <- fread("~/bigdata/alliston/HOBeSNP/data/serra_tables/12 TGFb_FDR 0_1.csv")
#tab <- fread("~/bigdata/alliston/HOBeSNP/data/serra_tables/2 Common DE Aging TGFb.csv")
#tab$Entrez_Id merges perfectly with the geneid used in txDB and the gene-based scores object
#get gene symbols
geneids <- as.data.frame(org.Hs.egSYMBOL)
geneids$gene_id <- as.integer(geneids$gene_id)

tab_gene <- merge(tab, gwas_gene, by.x = "Entrez_Id", by.y = "gene")
tab_gene <- merge(tab_gene, geneids, by.x = "Entrez_Id", by.y = "gene_id")

fwrite(tab_gene, file = "../results/table12_GWAS_frac.csv", na = "NA")

#there are 343 genes in Serra's table, but only 334 in annotation database
#how many Serra table genes are not found in annotation database?
#tab <- merge(tab, ghs, by.x = "Entrez_Id", by.y = "GENEID", all.x = T)
#gene id 199 is on chr 6, but it is absent from annotation file. I'll try a few other files later.

case_obs <- tab_gene[minG <= 0.0001 , .N]
control_obs <- tab_gene[minG > 0.0001, .N]
case_all <- gwas_gene[minG <= 0.0001, .N]
control_all <- gwas_gene[minG > 0.0001, .N]
rbind(c(case_obs, control_obs), c(case_all, control_all)) 
fisher.test( rbind(c(case_obs, control_obs), c(case_all, control_all)) )

```


